google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITJsonWriteReadTest.java

```diff
@@ -0,0 +1,156 @@
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.ParallelIntegrationTest;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.Value;
+import com.google.cloud.spanner.testing.EmulatorSpannerHelper;
+import com.google.cloud.spanner.testing.RemoteSpannerHelper;
+import com.google.common.io.Resources;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@Category(ParallelIntegrationTest.class)
+@RunWith(JUnit4.class)
+public class ITJsonWriteReadTest {
+
+  private static final String RESOURCES_DIR = "com/google/cloud/spanner/it/";
+  private static final String VALID_JSON_DIR = "valid";
+  private static final String INVALID_JSON_DIR = "invalid";
+
+  private static final String TABLE_NAME = "TestTable";
+
+  @ClassRule public static IntegrationTestEnv env = new IntegrationTestEnv();
+
+  private static DatabaseClient databaseClient;
+
+  @BeforeClass
+  public static void beforeClass() {
+    final RemoteSpannerHelper testHelper = env.getTestHelper();
+    final Database database =
+        testHelper.createTestDatabase(
+            "CREATE TABLE "
+                + TABLE_NAME
+                + "("
+                + "Id INT64 NOT NULL,"
+                + "json JSON"
+                + ") PRIMARY KEY (Id)");
+    databaseClient = testHelper.getDatabaseClient(database);
+  }
+
+  @Test
+  public void testWriteValidJsonValues() throws IOException {
+    assumeFalse("Emulator does not yet support JSON", EmulatorSpannerHelper.isUsingEmulator());
+
+    List<String> resources = getJsonFilePaths(RESOURCES_DIR + File.separator + VALID_JSON_DIR);
+
+    long id = 0L;
+    List<Mutation> mutations = new ArrayList<>();
+    List<String> jsons = new ArrayList<>();
+    for (String resource : resources) {
+      String jsonStr =
+          Resources.toString(
+              Resources.getResource(this.getClass(), VALID_JSON_DIR + File.separator + resource),
+              StandardCharsets.UTF_8);
+      jsons.add(jsonStr);
+      Mutation mutation =
+          Mutation.newInsertBuilder(TABLE_NAME)
+              .set("Id")
+              .to(id++)
+              .set("json")
+              .to(Value.json(jsonStr))
+              .build();
+      mutations.add(mutation);
+    }
+
+    databaseClient.write(mutations);
+    ResultSet resultSet =
+        databaseClient.singleUse().executeQuery(Statement.of("SELECT * FROM " + TABLE_NAME));
+    int count = 0;
+    while (resultSet.next()) {
+      count++;
```

Comment:
I think we should either (in order of preference):
1. Do a `resultSet.getJson("json")` here to verify that the value is also gettable in the client.
2. Or: Just re-write the query to a `SELECT COUNT(*) ...`

Reply:
I ended up going with option 2, because the JSON that gets returned is often not the same as what's inserted (e.g. `[123e45]` gets returned as `[1.23e+47]`), and it's difficult to automatically determine whether the returned value is equivalent to the inserted value without the same library that the backend uses to parse JSON. Since the tests are only there to confirm that the backend is able to accept the valid values, I don't think I'll bother spending the extra time manually adding the expected results for each file. Hopefully this is okay!